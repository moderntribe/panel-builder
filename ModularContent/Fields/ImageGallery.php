<?php


namespace ModularContent\Fields;

use ModularContent\AdminPreCache;
use ModularContent\Panel;


/**
 * Class ImageGallery
 *
 * @package ModularContent\Fields
 *
 * An image gallery field.
 *
 * get_panel_var() will return an array of attachment IDs
 */
class ImageGallery extends Field {

	protected $default = [];

	public function __construct( $args ) {
		$this->defaults[ 'strings' ] = [
			'button.edit_gallery' => __( 'Edit Gallery', 'modular-content' ),
		];
		parent::__construct( $args );
	}

	public function get_vars( $data, $panel ) {
		/*
		 * We have thumbnails stored with the IDs for quick loading in the admin,
		 * but we only want to send IDs to the front-end
		 */
		$images = parent::get_vars( $data, $panel );
		$ids = wp_list_pluck( $images, 'id' );

		$ids = apply_filters( 'panels_field_vars', $ids, $this, $panel );

		return $ids;
	}

	public function get_vars_for_api( $data, $panel ) {
		$image_ids = $this->get_vars( $data, $panel );
		$size = apply_filters( 'panels_image_gallery_field_size_for_api', 'full', $this, $data, $panel );
		$resources = [];

		foreach ( (array) $image_ids as $id ) {
			$resources[] = wp_get_attachment_image_src( $id, $size );
		}

		$resources = apply_filters( 'panels_field_vars_for_api', $resources, $data, $this, $panel );

		return $resources;
	}

	/**
	 * Add data relevant to this field to the precache
	 *
	 * @param mixed         $data
	 * @param AdminPreCache $cache
	 *
	 * @return void
	 */
	public function precache( $data, AdminPreCache $cache ) {
		$size = apply_filters( 'panels_image_gallery_field_size_for_admin', 'thumbnail', $this, $data );

		if ( is_array( $data ) ) {
			foreach ( $data as $image ) {
				$cache->add_image( $image[ 'id' ], $size );
			}
		}
	}

	/**
	 * Massage submitted data before it's saved.
	 *
	 * @param mixed $data
	 * @return array
	 */
	public function prepare_data_for_save( $data ) {
		$data = (array) $data;
		$cleaned = [];
		foreach ( $data as $image ) {
			if ( empty( $image[ 'id' ] ) ) {
				continue;
			}
			$thumbnail = wp_get_attachment_image_src( $image[ 'id' ], 'thumbnail' );
			if ( ! $thumbnail ) {
				continue;
			}
			$cleaned[] = [
				'id'        => (int) $image[ 'id' ],
				'thumbnail' => $thumbnail[ 0 ],
			];
		}
		return $cleaned;
	}
}