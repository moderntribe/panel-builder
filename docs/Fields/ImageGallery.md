## Image Gallery

An image gallery field.

get_panel_var() will return an array of attachment IDs


### Parameters

||||
|---|---|---|
| `label`       | `string` | The __() translated label for this field       |
| `name`        | `string` | The unique name for this field                 |
| `description` | `string` | The __() translated description for this field |

### Example

```php
$field = new ImageGallery( array(
    'label' => __('Gallery Images'),
    'name' => 'images',
    'description' => __( 'Images to display in the gallery' )
) );
```

### String Overrides

* `button.edit_gallery` - (`Edit Gallery`) - The Button Label
