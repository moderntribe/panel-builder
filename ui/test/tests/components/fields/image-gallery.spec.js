/* global expect b:true */

import React from 'react';
import { shallow, mount } from 'enzyme';
import ImageGallery from 'components/fields/image-gallery';
import renderer from 'react-test-renderer';

describe('Image Gallery field', () => {
	it('renders correctly', () => {
		const tree = renderer
			.create(<ImageGallery />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('renders as a <div>', () => {
		const wrapper = shallow(<ImageGallery />);
		expect(wrapper.type()).toEqual('div');
	});

	it('has a label', () => {
		const wrapper = shallow(<ImageGallery />);
		expect(wrapper.find('label.panel-field-label').length).toEqual(1);
	});

	it('has a gallery field name', () => {
		const wrapper = shallow(<ImageGallery />);
		expect(wrapper.find('input[name="gallery-field-name"]').length).toEqual(1);
	});

	it('has a gallery edit button', () => {
		const wrapper = shallow(<ImageGallery />);
		expect(wrapper.find('button').length).toEqual(1);
	});

	it('allows us to set and edit props', () => {
		const wrapper = mount(<ImageGallery description="This is a test description" label="Some test label" />);
		expect(wrapper.props().description).toEqual('This is a test description');
		wrapper.setProps({ description: 'Change description for test' });
		expect(wrapper.props().description).toEqual('Change description for test');

		expect(wrapper.props().label).toEqual('Some test label');
		wrapper.setProps({ label: 'Change label for test' });
		expect(wrapper.props().label).toEqual('Change label for test');
	});
});
