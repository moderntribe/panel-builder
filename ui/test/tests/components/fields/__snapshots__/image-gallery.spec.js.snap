exports[`Image Gallery field renders correctly 1`] = `
<div
  className="field panel-field">
  <label
    className="label panel-field-label">
    
  </label>
  <div
    data-label="Gallery"
    data-name="">
    <input
      name="gallery-field-name"
      type="hidden"
      value="" />
    <p
      className="galleryFieldControls">
      <button
        className="button button-large"
        onClick={[Function]} />
    </p>
    <div
      className="galleryFieldSelection"
      onClick={[Function]} />
  </div>
</div>
`;
